core/src/scene/sceneLoader.cpp

```diff
@@ -66,6 +73,266 @@ glm::vec3 parseVec3(const Node& node) {
+            shader.addSourceBlock("uniforms", "uniform " + name + " = " + value + ";");
+        }
+    }
+
+    Node blocksNode = shaders["blocks"];
+    if (blocksNode) {
+        for (const auto& block : blocksNode) {
+            std::string name = block.first.as<std::string>();
+            std::string value = block.second.as<std::string>();
+            shader.addSourceBlock(name, value); // TODO: Warn on unrecognized injection points
+        }
+    }
+
+}
+
+void SceneLoader::loadMaterial(YAML::Node matNode, Material& material, Scene& scene) {
+
+    Node diffuse = matNode["diffuse"];
+    if (diffuse) {
+        if (diffuse.IsMap()) {
+            material.setDiffuse(loadMaterialTexture(diffuse, scene));
+        } else if (diffuse.IsSequence()) {
+            material.setDiffuse(parseVec4(diffuse));
+        } else {
+            try {
+                float difValue = diffuse.as<float>();
+                material.setDiffuse(glm::vec4(difValue, difValue, difValue, 1.0));
+            } catch (const BadConversion& e) {
+                // TODO: css color parser and hex_values
+            }
+        }
+    }
+
+    Node ambient = matNode["ambient"];
+    if (ambient) {
+        if (ambient.IsMap()) {
+            material.setAmbient(loadMaterialTexture(ambient, scene));
+        } else if (ambient.IsSequence()) {
+            material.setAmbient(parseVec4(ambient));
+        } else {
+            try {
+                float ambientValue = ambient.as<float>();
+                material.setAmbient(glm::vec4(ambientValue, ambientValue, ambientValue, 1.0));
+            } catch (const BadConversion& e) {
+                // TODO: css color parser and hex_values
+            }
+        }
+    }
+
+    Node specular = matNode["specular"];
+    if (specular) {
+        if (specular.IsMap()) {
+            material.setSpecular(loadMaterialTexture(specular, scene));
+        } else if (specular.IsSequence()) {
+            material.setSpecular(parseVec4(specular));
+        } else {
+            try {
+                float specValue = specular.as<float>();
+                material.setSpecular(glm::vec4(specValue, specValue, specValue, 1.0));
+            } catch (const BadConversion& e) {
+                // TODO: css color parser and hex_values
+            }
+        }
+    }
+
+    Node shininess = matNode["shininess"];
+
+    if (shininess) {
+        try {
+            material.setShininess(shininess.as<float>());
+        } catch(const BadConversion& e) {
+            logMsg("Error: float value expected for shininess material parameter");
+        }
+    }
+
+    Node normal = matNode["normal"];
+    if (normal) {
+        material.setNormal(loadMaterialTexture(normal, scene));
+    }
+
+}
+
+MaterialTexture SceneLoader::loadMaterialTexture(YAML::Node matCompNode, Scene& scene) {
+
+    MaterialTexture matTex;
+
+    Node textureNode = matCompNode["texture"];
+    Node mappingNode = matCompNode["mapping"];
```

Comment:
texture [scale](https://github.com/tangrams/tangram-docs/blob/gh-pages/pages/textures.md#scale) and [amount](https://github.com/tangrams/tangram-docs/blob/gh-pages/pages/textures.md#amount) specification not handled here.